andela/octopus-bn-backend
# #170947443 Build new trip request notification feature

### What does this PR do?

- Create a controller to send notifications and connect it to trip rip request controllers
- Create Email template
- Send Email notifications
- Create Notifications table with a relationship to the user
- Update user migration
- Save notifications
- Setup socket.io
- Setup Static HTML page to test notifications
- Create an endpoint to view all notifications
- Create an endpoint to view all unread notifications
- Create an endpoint to change notification preferences
- Create a middleware to check if signed in user is a manager
- Add translations for error messages
- Create validations for inputs when the user is changing notification preferences
- Write tests
- Write documentation for this feature
### How should this be manually tested?

-  Clone the repo run the following commands:
`git checkout ft-new-trip-request-notification-170947450`
`npm i`
`npm run dev`
- Create a trip request and checkout `localhost:3000/api/v1/nottifications/static` to get an in-app notification the manager of the user who created the trip request should also get an email
- Use  GET `http://localhost:3000/api/v1/notifications/unread` to view unread notifications
- Use  GET `http://localhost:3000/api/v1/notifications` to view all notifications
- Use  PUT `http://localhost:3000/api/v1/notifications/preferences` to disable/enable in-App/Email notifications
### Implementation plan
[implementation plan](https://docs.google.com/document/d/1ArGkaHrd6TDcVMhF_mqinbn54aPDQTIUO0zrOGn0iJ8/edit?usp=sharing)

### Stories associated with PR

[#170947443](https://www.pivotaltracker.com/story/show/170947450)

### Screenshots if any
**Unread Notification**
<img width="1175" alt="Screen Shot 2020-02-27 at 05 23 02" src="https://user-images.githubusercontent.com/54870069/75409422-569acd80-5921-11ea-89c1-476ab6e61796.png">

**When there is no Notification**
<img width="1176" alt="Screen Shot 2020-02-27 at 05 24 10" src="https://user-images.githubusercontent.com/54870069/75409473-73cf9c00-5921-11ea-9b1f-5bf645adcfa8.png">
**Successfully updated user preferences on notifications**
<img width="1186" alt="Screen Shot 2020-02-27 at 05 25 13" src="https://user-images.githubusercontent.com/54870069/75409532-9feb1d00-5921-11ea-8285-4b4106f2109e.png">

**Email Notification**
<img width="691" alt="Screen Shot 2020-02-27 at 05 31 07" src="https://user-images.githubusercontent.com/54870069/75409897-a9c15000-5922-11ea-87fc-9ab0d2f9a4b6.png">

**In-App notifications(testing for the moment)**
<img width="1437" alt="Screen Shot 2020-02-27 at 13 07 32" src="https://user-images.githubusercontent.com/54870069/75439302-35a69c80-5962-11ea-88ce-cf8337753069.png">
